src/shared/image-policy.c

```diff
@@ -0,0 +1,653 @@
+        if (!policy)
+                return NULL;
+
+        return typesafe_bsearch(
+                        &(PartitionPolicy) { .designator = designator },
+                        ASSERT_PTR(policy)->policies,
+                        ASSERT_PTR(policy)->n_policies,
+                        partition_policy_compare);
+}
+
+static PartitionPolicyFlags partition_policy_normalized_flags(const PartitionPolicy *policy) {
+        PartitionPolicyFlags flags = ASSERT_PTR(policy)->flags;
+
+        /* If no protection flag is set, then this means all are set */
+        if ((flags & _PARTITION_POLICY_USE_MASK) == 0)
+                flags |= PARTITION_POLICY_OPEN;
+
+        /* If this is a verity or verity signature designator, then mask off all protection bits, this after
+         * all needs no protection, because it *is* the protection */
+        if (partition_verity_to_data(policy->designator) >= 0 ||
+            partition_verity_sig_to_data(policy->designator) >= 0)
+                flags &= ~(PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED|PARTITION_POLICY_ENCRYPTED);
+
+        /* if this designator has no verity concept, then mask off verity protection flags */
+        if (partition_verity_of(policy->designator) < 0)
+                flags &= ~(PARTITION_POLICY_VERITY|PARTITION_POLICY_SIGNED);
+
+        if ((flags & _PARTITION_POLICY_USE_MASK) == PARTITION_POLICY_ABSENT)
+                /* If the partition must be absent, then the gpt flags don't matter */
+                flags &= ~(_PARTITION_POLICY_READ_ONLY_MASK|_PARTITION_POLICY_GROWFS_MASK);
+        else {
+                /* If the gpt flags bits are not specified, set both options for each */
+                if ((flags & _PARTITION_POLICY_READ_ONLY_MASK) == 0)
+                        flags |= PARTITION_POLICY_READ_ONLY_ON|PARTITION_POLICY_READ_ONLY_OFF;
+                if ((flags & _PARTITION_POLICY_GROWFS_MASK) == 0)
+                        flags |= PARTITION_POLICY_GROWFS_ON|PARTITION_POLICY_GROWFS_OFF;
+        }
+
+        return flags;
+}
+
+PartitionPolicyFlags image_policy_get(const ImagePolicy *policy, PartitionDesignator designator) {
+        PartitionDesignator data_designator = _PARTITION_DESIGNATOR_INVALID;
+        PartitionPolicy *pp;
+
+        /* No policy means: everything may be used in any mode */
+        if (!policy)
+                return partition_policy_normalized_flags(
+                                &(const PartitionPolicy) {
+                                        .flags = PARTITION_POLICY_OPEN,
+                                        .designator = designator,
+                                });
+
+        pp = image_policy_bsearch(policy, designator);
+        if (pp)
+                return partition_policy_normalized_flags(pp);
+
+        /* Hmm, so this didn't work, then let's see if we can derive some policy from the underlying data
+         * partition in case of verity/signature partitions */
```

Comment:
well, for now there's little value in explicitly specifying a policy for verity or verity-sig partitions, we can derive a good policy from the policy on the data partitions anyway.

but it's not that there was no value: you can use that to declare a policy on the read-only/growfs flags if you like that is different as for the data partitions. That's not too useful, but if you press me hard I could come up with a case where this might even make sense: let's say that you prep an image that comes with verity and everything, but where you define a "developer" mode. once you enter that you "break the seal" and suddenly verity loses its effect and the normally read-only partition becomes writable. in such a case it might make sense to enforce that the growfs flags is enabled for the data partition (because once the seal is broken it might make sense to grow that data partition), but it should still be forced off for the verity partition then, because that data of course can never be grown. hence suddenly you want different policy for the gpt flags for the verity and for the data partition it protects, and thus the automatic synthesis is not going to be enough.

more importantly: the way i see it we should extend the policy language sooner or later. Specifically, one thing i'd like to do soon is encoding allowed file systems in the policy. i.e. that you can say root fs must be ext4 or btrfs, nothing else. (waiting for https://github.com/util-linux/util-linux/issues/1969 before we can implement this nicely, btw). for that I'd simply add a bunch of additional flags PARTITION_POLICY_EXT4,PARTITION_POLICY_BTRFS,PARTITION_POLICY_XFS and so on. now, this specific extension won't matter for the verity hash partition, but I think we might add other stuff where it eventually might make sense for example policy on more gpt part flags or so, or how uuids are generated for the partition or so.

it's also a lot simple to just treat all partitions types the same here, when enforcing them. and simply enforcement is secure enforcement...